test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java

```diff
@@ -0,0 +1,336 @@
+        return a * INT_MAX_MINUS_ONE + a; // a*MAX => a*(MIN-1) => a*MIN - a => (a<<31) - a
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.LSHIFT_I, "1" })
+    private static int mulAndAddToOverflow(int a) {
+        return a * Integer.MAX_VALUE + a; // a*(MAX+1) => a*(MIN) => a<<31
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.CON_I, "1" })
+    private static int mulAndAddToZero(int a) {
+        return a * -1 + a; // 0
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.LSHIFT_I, "1", IRNode.SUB_I, "1" })
+    private static int mulAndAddToMinus1(int a) {
+        return a * -2 + a; // a*-1 => a - (a<<1)
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.MUL_I, "1" })
+    private static int mulAndAddToMinus42(int a) {
+        return a * -43 + a; // a*-42
+    }
+
+    // --- long tests ---
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1" })
+    private static long mulAndAddToIntOverflowL(long a) {
+        return a * Integer.MAX_VALUE + a; // a*(INT_MAX+1)
+    }
+
+    private static final long LONG_MAX_MINUS_ONE = Long.MAX_VALUE - 1;
+
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1", IRNode.SUB_L, "1" })
+    private static long mulAndAddToMaxL(long a) {
+        return a * LONG_MAX_MINUS_ONE + a; // a*MAX => a*(MIN-1) => a*MIN - 1 => (a<<63) - 1
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1" })
+    private static long mulAndAddToOverflowL(long a) {
+        return a * Long.MAX_VALUE + a; // a*(MAX+1) => a*(MIN) => a<<63
+    }
+
+    // --- bit shift tests ---
+    @Test
+    @IR(failOn = {IRNode.ADD_I, IRNode.LSHIFT_I})
+    private static int bitShiftToOverflow() {
+        int i, x = 0;
+        for (i = 0; i < 32; i++) {
+            x = i;
+        }
+
+        // x = 31 (phi), i = 32 (phi + 1)
+        return i + (x << i) + i; // Expects 32 + 31 + 32 = 95
+    }
+
+    @Test
+    @IR(failOn = {IRNode.ADD_L, IRNode.LSHIFT_L})
+    private static long bitShiftToOverflowL() {
+        int i, x = 0;
+        for (i = 0; i < 64; i++) {
+            x = i;
+        }
+
+        // x = 63 (phi), i = 64 (phi + 1)
+        return i + (x << i) + i; // Expects 64 + 63 + 64 = 191
+    }
+
+    // --- random tests ---
+    private static final int CON1_I, CON2_I, CON3_I, CON4_I;
+    private static final long CON1_L, CON2_L, CON3_L, CON4_L;
+
+    static {
+        RestrictableGenerator<Integer> genI = Generators.G.powerOfTwoInts(16);
+        CON1_I = genI.next();
+        CON2_I = genI.next();
+        CON3_I = genI.next();
+        CON4_I = genI.next();
+
+        RestrictableGenerator<Long> genL = Generators.G.powerOfTwoLongs(16);
+        CON1_L = genL.next();
+        CON2_L = genL.next();
+        CON3_L = genL.next();
+        CON4_L = genL.next();
+    }
```

Comment:
Is there a reason why you are restricting the values to `powerOfTwoLongs`? I think it would be better if you just take the most general generator.
```
private static final RestrictableGenerator<Integer> GEN_INT = Generators.G.ints();
private static final RestrictableGenerator<Long> GEN_LONG = Generators.G.longs();
```

Reply:
Sorry misunderstood your previous comment. I thought you wanted to test especially power-of-two's. Now I understand you meant more "interesting" randoms (potentially close to power-of-two's)